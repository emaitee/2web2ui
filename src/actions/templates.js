import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';

export function listTemplates() {
  return sparkpostApiRequest({
    type: 'LIST_TEMPLATES',
    meta: {
      method: 'GET',
      url: '/templates'
    }
  });
}

export function getDraft(id) {
  return sparkpostApiRequest({
    type: 'GET_DRAFT_TEMPLATE',
    meta: {
      method: 'GET',
      url: `/templates/${id}`,
      params: {
        draft: true
      }
    }
  });
}

export function getPublished(id) {
  return sparkpostApiRequest({
    type: 'GET_PUBLISHED_TEMPLATE',
    meta: {
      method: 'GET',
      url: `/templates/${id}`,
      params: {
        draft: false
      }
    }
  });
}

export function create(data) {
  return sparkpostApiRequest({
    type: 'CREATE_TEMPLATE',
    meta: {
      method: 'POST',
      url: '/templates',
      data
    }
  });
}

export function update(data, params = {}) {
  const id = data.id;
  return sparkpostApiRequest({
    type: 'UPDATE_TEMPLATE',
    meta: {
      method: 'PUT',
      url: `/templates/${id}`,
      data,
      params: { ...params }
    }
  });
}

export function publish(id) {
  return sparkpostApiRequest({
    type: 'PUBLISH_TEMPLATE',
    meta: {
      method: 'PUT',
      url: `/templates/${id}`,
      data: { published: true }
    }
  });
}

export function deleteTemplate(id) {
  return sparkpostApiRequest({
    type: 'DELETE_TEMPLATE',
    meta: {
      method: 'DELETE',
      url: `/templates/${id}`
    }
  });
}
