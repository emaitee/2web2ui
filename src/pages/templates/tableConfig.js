import React from 'react';
import { Link } from 'react-router-dom';
import { format } from 'date-fns';

export const getRowData = ({ published, id, name, last_update_time, subaccount_id, shared_with_subaccounts }) => [
  <Link to={`/templates/edit/${id}`}>{name}</Link>,
  id,
  published ? 'Published' : 'Draft',
  format(last_update_time, 'MMM D, YYYY [at] h:mma'),
  shared_with_subaccounts ? 'All' : subaccount_id
];

export const columns = [
  { label: 'Name', width: '25%' },
  { label: 'ID', width: '25%' },
  { label: 'Status', width: '15%' },
  { label: 'Updated' },
  { label: 'Subaccount', width: '15%' }
];

export const filterBoxConfig = {
  show: true,
  exampleModifiers: ['id', 'name'],
  compareKeys: ['name', 'id', 'subaccount_id']
};
