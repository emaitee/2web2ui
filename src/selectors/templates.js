/* eslint-disable */
import { createSelector } from 'reselect';
import { formValues } from 'redux-form';
import sortMatch from 'src/helpers/sortMatch';

export const templatesListSelector = (state) => state.templates.list;
export const templateByIdSelector = (state, props) => state.templates.byId[props.match.params.id] || { draft: {}, published: {}};

const filtersSelector = (state) => state.form.templateFilters && state.form.templateFilters.values;

export const filteredTemplatesSelector = createSelector(
  templatesListSelector, filtersSelector,
  (list, filters = {}) => {
    let matches = list;
    const { search = false, status = {}, subaccount = {} } = filters;

    if (search) {
      matches = sortMatch(
        list, search,
        (item) => `${item.name} ID: ${item.id}`
      );
    }

    // console.log(matches)
    return matches;
  }
);
