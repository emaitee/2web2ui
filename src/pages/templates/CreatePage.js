import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { reduxForm, formValueSelector } from 'redux-form';
// Actions
import { create } from '../../actions/templates';

// Components
import { Layout } from 'src/components';
import Form from './components/Form';
import Editor from './components/Editor';
import { Page, Grid } from '@sparkpost/matchbox';

const FORM_NAME = 'templateCreate';

class CreatePage extends Component {
  state = {
    shouldRedirect: false
  }

  handleCreate(values) {
    const { create } = this.props;
    return create(values)
      .then(() => this.setState({ shouldRedirect: true }));
  }

  renderPageHeader() {
    const { handleSubmit, submitting } = this.props;

    const primaryAction = {
      content: 'Save Template',
      onClick: handleSubmit((values) => this.handleCreate(values)),
      disabled: submitting
    };

    const backAction = {
      content: 'Templates',
      Component: Link,
      to: '/templates'
    };

    return (
      <Page
        primaryAction={primaryAction}
        breadcrumbAction={backAction}
        title='New Template'
      />
    );
  }

  render() {
    const { id } = this.props;

    if (this.state.shouldRedirect) {
      return <Redirect to={`/templates/edit/${id}`} />;
    }

    return (
      <Layout.App>
        { this.renderPageHeader() }
        <Grid>
          <Grid.Column xs={12} lg={4}>
            <Form newTemplate={true} name={FORM_NAME} />
          </Grid.Column>
          <Grid.Column xs={12} lg={8}>
            <Editor name={FORM_NAME} />
          </Grid.Column>
        </Grid>
      </Layout.App>
    );
  }
}

const selector = formValueSelector(FORM_NAME);
const mapStateToProps = (state) => ({
  id: selector(state, 'id')
});

const formOptions = {
  form: FORM_NAME
};

export default connect(mapStateToProps, { create })(reduxForm(formOptions)(CreatePage));
