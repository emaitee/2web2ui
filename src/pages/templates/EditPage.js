/* eslint max-lines: ["error", 200] */
import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';
// Actions
import { getDraft, getPublished, update, deleteTemplate, publish } from '../../actions/templates';
import { showAlert } from 'src/actions/globalAlert';

// Selectors
import { getTemplateById } from 'src/selectors/templates';

// Components
import { Layout } from 'src/components';
import Form from './components/Form';
import Editor from './components/Editor';
import DeleteModal from './components/DeleteModal';
import { Page, Grid } from '@sparkpost/matchbox';

const FORM_NAME = 'templateEdit';

class EditPage extends Component {
  state = {
    redirectTo: null,
    deleteOpen: false
  };

  componentDidMount() {
    const { match, getDraft, getPublished } = this.props;
    // Do nothing if these fail
    getDraft(match.params.id).catch((err) => err);
    getPublished(match.params.id).catch((err) => err);
  }

  handlePublish(values) {
    const { publish, match, showAlert } = this.props;
    return publish(match.params.id)
      .then(() => this.setState({ redirectTo: `/templates/edit/${match.params.id}/published` }))
      .then(() => showAlert({ type: 'success', message: 'Template published' }))
      .catch((err) => showAlert({ type: 'error', message: 'Could not publish template', details: err.message }));
  }

  handleSave(values) {
    const { update, match, getDraft, showAlert } = this.props;
    return update(values)
      .then(() => getDraft(match.params.id))
      .then(() => showAlert({ type: 'success', message: 'Template saved' }))
      .catch((err) => showAlert({ type: 'error', message: 'Could not save template', details: err.message }));
  }

  handleDelete() {
    const { deleteTemplate, match, showAlert } = this.props;
    return deleteTemplate(match.params.id)
      .then(() => this.setState({ redirectTo: '/templates/' }))
      .then(() => showAlert({ message: 'Template deleted' }))
      .catch((err) => showAlert({ type: 'error', message: 'Could not delete template', details: err.message }));
  }

  handleDeleteModalToggle() {
    this.setState({ deleteOpen: !this.state.deleteOpen });
  }

  componentDidUpdate() {
    const { loading, template, showAlert } = this.props;

    if (!loading && !Object.keys(template.draft).length && !Object.keys(template.published).length) {
      this.setState({ redirectTo: '/templates/' }); // Redirect if no draft or published found
      showAlert({ type: 'error', message: 'Could not find template' });
    }
  }

  renderPageHeader() {
    const {
      handleSubmit,
      template,
      match,
      submitting
    } = this.props;
    const published = template.published;

    const primaryAction = {
      content: 'Publish Template',
      onClick: handleSubmit((values) => this.handlePublish(values)),
      disabled: submitting
    };

    const viewActions = published ? [
      {
        content: 'View Published',
        Component: Link,
        to: `/templates/edit/${match.params.id}/published`
      }
    ] : [];

    const secondaryActions = [
      ...viewActions,
      {
        content: 'Save as Draft',
        onClick: handleSubmit((values) => this.handleSave(values)),
        disabled: submitting
      },
      { content: 'Delete', onClick: () => this.handleDeleteModalToggle() },
      { content: 'Duplicate', disabled: true },
      { content: 'Preview & Send', disabled: true }
    ];

    const backAction = {
      content: 'Templates',
      Component: Link,
      to: '/templates'
    };

    return (
      <Page
        primaryAction={primaryAction}
        secondaryActions={secondaryActions}
        breadcrumbAction={backAction}
        title={`${match.params.id} (Draft)`}
      />
    );
  }

  render() {
    const { loading } = this.props;

    if (this.state.redirectTo) {
      return <Redirect to={this.state.redirectTo} />;
    }

    return (
      <Layout.App loading={loading}>
        { this.renderPageHeader() }
        <Grid>
          <Grid.Column xs={12} lg={4}>
            <Form name={FORM_NAME} />
          </Grid.Column>
          <Grid.Column xs={12} lg={8}>
            <Editor name={FORM_NAME} />
          </Grid.Column>
        </Grid>
        <DeleteModal
          open={this.state.deleteOpen}
          handleToggle={() => this.handleDeleteModalToggle()}
          handleDelete={() => this.handleDelete()}/>
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => {
  const template = getTemplateById(state, props);
  return {
    loading: state.templates.getLoading,
    template,
    // For templates with published but no draft, pull in published values
    initialValues: template.draft || template.published
  };
};
const formOptions = {
  form: FORM_NAME,
  enableReinitialize: true // required to update initial values from redux state
};

export default connect(mapStateToProps, {
  getDraft,
  getPublished,
  update,
  deleteTemplate,
  publish,
  showAlert
})(reduxForm(formOptions)(EditPage));
