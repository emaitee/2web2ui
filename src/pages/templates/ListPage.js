import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import config from 'src/config';

// Actions
import { listTemplates } from 'src/actions/templates';
import { getTemplates, getfilterTemplates } from 'src/selectors/templates';

// Components
import { Layout, TableCollection, ApiErrorBanner } from 'src/components';
import { Page, EmptyState } from '@sparkpost/matchbox';
import Filters from './components/Filters';
import { getRowData, TableHeader } from './components/TableComponents';

const primaryAction = {
  content: 'Create Template',
  to: '/templates/create',
  Component: Link
};

export class ListPage extends Component {

  componentDidMount() {
    this.props.listTemplates();
  }

  renderError() {
    return (
      <ApiErrorBanner
        message={'Sorry, we seem to have had some trouble loading your templates.'}
        errorDetails={this.props.error.message}
        reload={this.props.listTemplates}
      />
    );
  }

  renderCollection() {
    return (
      <TableCollection
        headerComponent={<TableHeader />}
        rows={this.props.templates}
        getRowData={getRowData}
        pagination
      />
    );
  }

  render() {
    const { count, loading, error } = this.props;

    if (!loading && count === 0) {
      return (
        <Layout.App>
          <EmptyState
            title='Manage your email templates'
            action={primaryAction} >
            {/* secondaryAction={content: 'Learn More', onClick: Learn_More()} > */}
            <p>Build, test, preview and send your transmissions.</p>
          </EmptyState>
        </Layout.App>
      );
    }

    return (
      <Layout.App loading={loading}>
        <Page
          primaryAction={primaryAction}
          title='Templates'
        />
        {error && this.renderError()}
        {!error && count > config.filters.minCount && <Filters />}
        {!error && this.renderCollection()}
      </Layout.App>
    );
  }
}

function mapStateToProps(state) {
  return {
    count: getTemplates(state).length,
    templates: getfilterTemplates(state),
    loading: state.templates.listLoading,
    error: state.templates.listError
  };
}

export default withRouter(connect(mapStateToProps, { listTemplates })(ListPage));
