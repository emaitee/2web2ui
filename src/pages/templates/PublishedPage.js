import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';

// Actions
import { getPublished } from '../../actions/templates';

// Selectors
import { getTemplateById } from 'src/selectors/templates';

// Components
import { Layout } from 'src/components';
import Form from './components/Form';
import Editor from './components/Editor';
import { Page, Grid } from '@sparkpost/matchbox';

const FORM_NAME = 'templatePublished';

class PublishedPage extends Component {
  state = {
    shouldRedirectToPublished: false
  };

  componentWillMount() {
    const { match, getPublished } = this.props;
    getPublished(match.params.id);
  }

  renderPageHeader() {
    const { match } = this.props;

    const secondaryActions = [
      {
        content: 'View Draft',
        Component: Link,
        to: `/templates/edit/${match.params.id}`
      },
      {
        content: 'Preview & Send',
        disabled: true
      }
    ];

    const backAction = {
      content: 'Templates',
      Component: Link,
      to: '/templates'
    };

    return (
      <Page
        secondaryActions={secondaryActions}
        breadcrumbAction={backAction}
        title={`${match.params.id} (Published)`}
      />
    );
  }

  render() {
    const { loading } = this.props;

    return (
      <Layout.App loading={loading}>
        { this.renderPageHeader() }
        <Grid>
          <Grid.Column xs={12} lg={4}>
            <Form name={FORM_NAME} published />
          </Grid.Column>
          <Grid.Column xs={12} lg={8}>
            <Editor name={FORM_NAME} published />
          </Grid.Column>
        </Grid>
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => ({
  loading: state.templates.getLoading,
  initialValues: getTemplateById(state, props).published
});

const formOptions = {
  form: FORM_NAME,
  enableReinitialize: true // required to update initial values from redux state
};

export default connect(mapStateToProps, { getPublished })(reduxForm(formOptions)(PublishedPage));
