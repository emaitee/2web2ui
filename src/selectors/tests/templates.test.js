import { getfilterTemplates, getTemplateById } from '../templates';

describe('Templates by id Selector', () => {
  const store = {
    templates: {
      byId: {
        Ape: {
          name: 'Ape',
        }
      }
    }
  };

  it('returns template', () => {
    const props = { match: { params: { id: 'Ape' }}};
    expect(getTemplateById(store, props)).toMatchSnapshot();
  });

  it('returns empty draft and published', () => {
    const props = { match: { params: { id: 'Nope' }}};
    expect(getTemplateById(store, props)).toMatchSnapshot();
  });
});

describe('Filter Templates Selector', () => {
  const store = {
    templates: {
      list: [
        {
          name: 'Zebra',
          published: true,
          subaccount_id: '1'
        },
        {
          name: 'Ape',
          published: false
        },
        {
          name: 'shared',
          published: false,
          shared_with_subaccounts: true
        },
      ]
    },
    form: {
      templateFilters: {}
    }
  };

  it('returns unfiltered list', () => {
    expect(getfilterTemplates(store)).toMatchSnapshot();
  });

  it('returns searched list', () => {
    const search = { form: { templateFilters: { values: { search: 'ebra' }}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });

  it('filters draft templates', () => {
    const search = { form: { templateFilters: { values: { status: { draft: true }}}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });

  it('filters published templates', () => {
    const search = { form: { templateFilters: { values: { status: { published: true }}}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });

  it('filters subaccount assigned', () => {
    const search = { form: { templateFilters: { values: { subaccount: { subaccount: true }}}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });

  it('filters subaccount all', () => {
    const search = { form: { templateFilters: { values: { subaccount: { all: true }}}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });

  it('filters subaccount master', () => {
    const search = { form: { templateFilters: { values: { subaccount: { master: true }}}}};
    expect(getfilterTemplates({ ...store, ...search})).toMatchSnapshot();
  });
});
