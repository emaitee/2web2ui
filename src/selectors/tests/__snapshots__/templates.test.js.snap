// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Templates Selector filters draft templates 1`] = `
Array [
  Object {
    "name": "Ape",
    "published": false,
  },
  Object {
    "name": "shared",
    "published": false,
    "shared_with_subaccounts": true,
  },
]
`;

exports[`Filter Templates Selector filters published templates 1`] = `
Array [
  Object {
    "name": "Zebra",
    "published": true,
    "subaccount_id": "1",
  },
]
`;

exports[`Filter Templates Selector filters subaccount all 1`] = `
Array [
  Object {
    "name": "shared",
    "published": false,
    "shared_with_subaccounts": true,
  },
]
`;

exports[`Filter Templates Selector filters subaccount assigned 1`] = `
Array [
  Object {
    "name": "Zebra",
    "published": true,
    "subaccount_id": "1",
  },
]
`;

exports[`Filter Templates Selector filters subaccount master 1`] = `
Array [
  Object {
    "name": "Ape",
    "published": false,
  },
]
`;

exports[`Filter Templates Selector returns searched list 1`] = `
Array [
  Object {
    "name": "Zebra",
    "published": true,
    "subaccount_id": "1",
  },
]
`;

exports[`Filter Templates Selector returns unfiltered list 1`] = `
Array [
  Object {
    "name": "Zebra",
    "published": true,
    "subaccount_id": "1",
  },
  Object {
    "name": "Ape",
    "published": false,
  },
  Object {
    "name": "shared",
    "published": false,
    "shared_with_subaccounts": true,
  },
]
`;

exports[`Templates by id Selector returns empty draft and published 1`] = `
Object {
  "draft": Object {},
  "published": Object {},
}
`;

exports[`Templates by id Selector returns template 1`] = `
Object {
  "name": "Ape",
}
`;
